### ISSUE TITLE
Naming for abilities lua and wml objects.

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#7932]
I still want to add access abilities (and specials) from lua and formula via abilitiy objects (as part of #7589 (outdated)).

One of the problems i faced is that the natural names for that is already taken (unit.abilities returns a list of strings, weapon.specials returns a list of config objects).

So i opened this issue to figure out what the best approach is (basicialy whether to deprecate those to free these names to returns an actual list of ability objects instead, or what else to do)

**Comment by Pentarctagon** [comment]
Are `unit.abilities` and `weapon.specials` the only two?

**Comment by CelticMinstrel** [comment]
Would it work to change `unit.abilities` to return a list of ability objects _and_ make ability objects have a `__tostring` and `__cat` method so that they behave like strings if something tries to use them like one? We'd probably also add a deprecation message to those functions to recommend that people fetch the ability's name directly as well.

> Are `unit.abilities` and `weapon.specials` the only two?

I don't have examples but I bet there are more. Also, I'll note that this problem affects WFL as well, not just Lua.

**Comment by gfgtdf** [comment]
> Are `unit.abilities` and `weapon.specials` the only two?

What celmin said, the same two things exists in formula too (and return a list of ids there).

> Would it work to change unit.abilities to return a list of ability objects and make ability objects have a __tostring and __cat method so that they behave like strings if something tries to use them like one? We'd probably also add a deprecation message to those functions to recommend that people fetch the ability's name directly as well.

This would probably only work in some rare cases, i think most codes use it for tests like `for i,v in ipairs(unit.abilities) ido s_leadership = v =="leadership" ..` and don't call __tostring directly (maybe we should overwrite __eq though , but that coudl cause trouble elsewhere), also idk whether  this approach applies to formula aswell.

> That doesn't work so well for weapon.specials, but I think most uses of that would just continue to work if you change it to an ability object, wouldn't they? Subtags are the one thing that would break, but I can't think of any reason someone would want to access a special's subtags.

hmm not sure, i'd actually be surprised if  `weapon.specials` is currently used anywhere at all, but i might be wrong.

**Comment by Pentarctagon** [comment]
If we have a list of what would be affected, then the they can be searched for on the add-ons server for whether they're used or not.

**Comment by CelticMinstrel** [comment]
> > Would it work to change unit.abilities to return a list of ability objects and make ability objects have a __tostring and __cat method so that they behave like strings if something tries to use them like one? We'd probably also add a deprecation message to those functions to recommend that people fetch the ability's name directly as well.
> 
> This would probably only work in some rare cases, i think most codes use it for tests like `for i,v in ipairs(unit.abilities) ido s_leadership = v =="leadership" ..` and don't call __tostring directly (maybe we should overwrite __eq though , but that coudl cause trouble elsewhere), also idk whether this approach applies to formula aswell.

Yeah, I'd probably overwrite `__eq` as well like you mentioned. Note that these metamethods would be transitional and exist primarily to give a deprecation message so people update their code, and to let their code continue to function until they do update it. So if it causes problems when people _intentionally_ rely in them, that's not a big deal in my opinion.

But you're right that the approach may not really apply to WFL. I _think_WFL does offer a way to override `=` though, so maybe that could help.
